Installation issue: openblas

### Steps to reproduce the issue

```console
spack install openblas %clang @15
```


### Error message

The reported error is
```console
linktest.c:2:1: error: call to undeclared function 'lsame_'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
```

See https://github.com/xianyi/OpenBLAS/issues/3760 for details.


### Information on your system

The error is caused by Clang 15, which does not allow implicit function declarations any more. I have a patch.

### Additional information

_No response_

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/openblas/linktest.patch
+++ b/var/spack/repos/builtin/packages/openblas/linktest.patch
@@ -1,0 +1,21 @@
+--- a/exports/gensymbol
++++ b/exports/gensymbol
+@@ -3914,6 +3914,18 @@
+     @underscore_objs = (@underscore_objs, @misc_common_objs);
+     @no_underscore_objs = (@no_underscore_objs, @misc_common_objs);
+ 
++    foreach $objs (@underscore_objs) {
++        print "void ", $symbolprefix, $objs, $bu, $symbolsuffix, "();\n" if $objs ne "xerbla";
++    }
++
++    foreach $objs (@need_2underscore_objs) {
++        print "void ", $symbolprefix, $objs, $bu, $bu, $symbolsuffix, "();\n";
++    }
++
++    foreach $objs (@no_underscore_objs) {
++        print "void ", $symbolprefix, $objs, $symbolsuffix, "();\n";
++    }
++
+     print "int main(void){\n";
+     foreach $objs (@underscore_objs) {
+         print $symbolprefix, $objs, $bu, $symbolsuffix, "();\n" if $objs ne "xerbla";

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -153,6 +153,10 @@ class Openblas(MakefilePackage):
     # Use /usr/bin/env perl in build scripts
     patch("0001-use-usr-bin-env-perl.patch", when="@:0.3.13")
 
+    # Declare external functions in linktest
+    # See <https://github.com/xianyi/OpenBLAS/issues/3760>
+    patch("linktest.patch", when="@0.3.20")
+
     # See https://github.com/spack/spack/issues/19932#issuecomment-733452619
     conflicts("%gcc@7.0.0:7.3,8.0.0:8.2", when="@0.3.11:")
 

